feat(noUnusedFunctionParameters): add lint for unused function parameters, instead of using noUnusedVariables

printfn opened the pull request:
## Summary

This adds a new lint `noUnusedFunctionParameters` that detects unused parameters. Previously these were detected by `noUnusedVariables`.

## Test Plan

I added tests for the new lint.

printfn commented:
Rebased and fixed merge conflicts. Would someone be able to review this?

ematipico commented:
What's the reasoning that drove you to implement the option? This reasoning will be very useful in explaining to the users why this option exists and how they can use it in their code.

printfn commented:
I'm using biome on a large legacy codebase, slowly migrating from ESLint. Having the option to enable `noUnusedVariables` would be super helpful in improving code quality, but there are *thousands* of unused arguments that shouldn't have to meaninglessly be renamed to have an underscore in order to still benefit from linting for unused variables.

I think that having unused arguments is much less of a code smell than entirely unused variables. In fact, especially when creating and passing around lambda functions, having unused arguments present can be really useful to readers so they immediately know that another argument is available if the code needs to be modified in the future.

While prefixing arguments with an underscore is still an option, in my case this would require renaming thousands of parameters and introducing a non-trivial amount of line noise.

One pattern that frequently comes up is:

```typescript
foo((a, b, c) => b);
```

In code like this, having the option to provide useful argument names without having to prefix everything is very useful.

Common examples:

```typescript
return new Promise((accept, reject) => { ... }); // not using `reject`

let arr = loadArr(); // `arr` has type `[string, string][]`
arr.filter(([k, v]) => v.length > 0); // `k` is unused

componentDidUpdate(prevProps: Props, prevState: State) { ... } // not using `prevState`
```

While in many of the examples it might make sense to remove the extra argument or add an underscore, doing this thousands of times across a legacy codebase is unreasonable and shouldn't be necessary to still benefit from `noUnusedVariables`.

printfn commented:
It's also worth pointing out that there are many repos on GitHub that set ESLint's `args` to `none`:

https://github.com/search?q=%2F%22no-unused-vars%22%3A%5Cs*%5C%5B%5Cs*%22%28warn%7Cerror%29%22%2C%5Cs*%5C%7B.*%22args%22%3A%5Cs*%22none%22.*%5C%7D%5C%5D%2F&type=code

printfn commented:
Alright, pretty sure I got it all working now. I've added a new nursery lint `noUnusedFunctionParameters` and changed `noUnusedVariables` to no longer include function parameters.

ematipico commented on crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs:
```diff
@@ -173,21 +151,7 @@ fn suggested_fix_if_unused(binding: &AnyJsIdentifierBinding) -> Option<Suggested
         AnyJsBindingDeclaration::JsArrowFunctionExpression(_) => {
             suggestion_for_binding(binding)
         }
-        AnyJsBindingDeclaration::TsPropertyParameter(_) => None,
-        AnyJsBindingDeclaration::JsFormalParameter(parameter) => {
-            if is_function_that_is_ok_parameter_not_be_used(&parameter.parent_function()) {
-                None
-            } else {
-                suggestion_for_binding(binding)
-            }
-        }
-        AnyJsBindingDeclaration::JsRestParameter(parameter) => {
-            if is_function_that_is_ok_parameter_not_be_used(&parameter.parent_function()) {
-                None
-            } else {
-                suggestion_for_binding(binding)
-            }
-        }
```
This changes the behaviour of the rule, and it's a breaking change. We should change this only when the new rule goes out of nursery.

printfn replied:
I've reverted the changes to `noUnusedVariables` now

ematipico commented on crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs:
```diff
@@ -26,9 +24,11 @@ declare_rule! {
+    /// This rule won't report unused imports or unused function parameters.
     /// If you want to report unused imports,
-    /// enable [noUnusedImports](https://biomejs.dev/linter/rules/no-unused-imports/).
+    /// enable [noUnusedImports](https://biomejs.dev/linter/rules/no-unused-imports/);
+    /// to report unused function parameters, enable
+    /// [noUnusedFunctionParameters](https://biomejs.dev/linter/rules/no-unused-function-parameters/)
```
We can't change the documentation like this. This change will go straight to production and users will be misled. I think we should **add** a new paragraph like this:

```md
From `v1.9.0`, the rule won't check unused function parameters anymore, users should switch to [noUnusedFunctionParameters](https://biomejs.dev/linter/rules/no-unused-function-parameters/)
```

I wrote `1.9.0` because in `v1.8.0` the new rule is nursery, and we can't suggest a rule that is still in their incubation period.

printfn replied:
Updated docs as suggested

printfn commented:
this is ready for another review @ematipico
